qlib/data/storage/storage.py

```diff
@@ -0,0 +1,420 @@
+        raise NotImplementedError("Subclass of BaseStorage must implement `check_exists` method")
+
+
+class UserInstrumentStorage(InstrumentStorage):
+
+    @property
+    def data(self) -> Dict[InstKT, InstVT]:
+        '''get all data'''
+        raise NotImplementedError("Subclass of InstrumentStorage must implement `data` method")
+
+    def check_exists(self) -> bool:
+        '''check if storage(uri) exists, if not exists: return False'''
+        raise NotImplementedError("Subclass of BaseStorage must implement `check_exists` method")
+
+
+class UserFeatureStorage(FeatureStorage):
+
+    def __getitem__(self, s: slice) -> pd.Series:
+        '''x.__getitem__(slice(start: int, stop: int, step: int)) <==> x[start:stop:step]
+
+        Returns
+        -------
+            pd.Series(values, index=pd.RangeIndex(start, len(values))
+        '''
+        raise NotImplementedError(
+            "Subclass of FeatureStorage must implement `__getitem__(s: slice)` method"
+        )
+
+    def check_exists(self) -> bool:
+        '''check if storage(uri) exists, if not exists: return False'''
+        raise NotImplementedError("Subclass of BaseStorage must implement `check_exists` method")
+
+"""
+
+
+class StorageMeta(type):
+    """unified management of raise when storage is not exists"""
+
+    def __new__(cls, name, bases, dict):
+        class_obj = type.__new__(cls, name, bases, dict)
+
+        # The calls to __iter__ and __getitem__ do not pass through __getattribute__.
+        # In order to throw an exception before calling __getitem__, use the metaclass
+        _getitem_func = getattr(class_obj, "__getitem__")
+
+        def _getitem(obj, item):
+            getattr(obj, "_check")()
+            try:
+                res = _getitem_func(obj, item)
+            except Exception as e:
+                raise ValueError(f"{obj.raise_info}\n\tStorage exception info: {str(e)}")
+            return res
+
+        setattr(class_obj, "__getitem__", _getitem)
+        return class_obj
+
+
+class BaseStorage(metaclass=StorageMeta):
+    @property
+    def storage_name(self) -> str:
+        return re.findall("[A-Z][^A-Z]*", self.__class__.__name__)[-2].lower()
+
+    @property
+    def raise_info(self):
+        parameters_info = [
+            f"{_k}={_v}"
+            for _k, _v in self.__dict__.items()
+            if not isinstance(_v, (dict,)) or (hasattr(_v, "__len__") and len(_v) < 3)
+        ]
+        return f"{self.storage_name.lower()} not exists, storage parameters: {parameters_info}"
+
+    def check_exists(self) -> bool:
+        """check if storage(uri) exists, if not exists: return False"""
+        raise NotImplementedError("Subclass of BaseStorage must implement `check_exists` method")
+
+    def clear(self) -> None:
+        """clear storage"""
+        raise NotImplementedError("Subclass of BaseStorage must implement `clear` method")
+
+    def __len__(self) -> 0:
+        return len(self.data) if self.check_exists() else 0
```

Comment:
`self.data` is not a method or member of the interface

Reply:
Besides, please give documents about this implementation to tell users that this is not an optimal implementation in many cases.